### PR TITLE
Check datadirectory owner, not config owner.

### PR DESCRIPTION
**Pull request by Dreamsorcerer**
Changes permission checks to use datadirectory, rather than config file.
Step 1 towards root-owned configs (#11075).

For root-owned configs, the only thing that needs to be writeable is the data directory. For non-root configs, this check shouldn't make any difference, it's very likely the same user is used for both config and data dir.

Currently, the user shouldn't see any difference in behaviour, as the check here would still disallow a root-owned config:
https://github.com/nextcloud/server/blob/master/lib/base.php#L247
Once the rest of the work to support root-owned configs is complete, then this check should be removed (`|| !$configFileWritable && \OCP\Util::needUpgrade()`).

### CODE REVIEW

**Comment by szaimen** [comment]
cc @skjnldsv, @eneiluj, @icewind1991 and @ArtificialOwl for feedback on this

**Comment by skjnldsv** [comment]
@Dreamsorcerer could you rebase and resolve conflicts? :)
Thanks a lot for your interest in that feature :bow:

**Comment by Dreamsorcerer** [comment]
Updated

**Review by skjnldsv — CHANGES_REQUESTED** [review]

File: console.php [hunk 1]
```diff
@@ -59,12 +59,14 @@ function exceptionHandler($exception) {
...
+
 	if (!function_exists('posix_getuid')) {
 		echo "The posix extensions are required - see https://www.php.net/manual/en/book.posix.php" . PHP_EOL;
 		exit(1);
 	}
 	$user = posix_getuid();
-	$configUser = fileowner(OC::$configDir . 'config.php');
+	$configUser = fileowner($config->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data'));
 	if ($user !== $configUser) {
 		echo "Console has to be executed with the user that owns the file config/config.php" . PHP_EOL;
```
[the remark this answers is not in the document]
Comment (Dreamsorcerer) [reply console.php:71]:
```suggestion
		echo "Console has to be executed with the user that owns the data directory" . PHP_EOL;
```

**Review by skjnldsv — APPROVED** [review]
I would prefer better commits messages 🙈
But I'm okay with the code 👍

**Comment by Dreamsorcerer** [comment]
> I would prefer better commits messages 🙈 But I'm okay with the code 👍

Sure, but it's a trivial change. I assume it'll be squashed.

**Comment by skjnldsv** [comment]
No, we do not squash, we want to preserve the commits history

File: cron.php [hunk 2]
```diff
@@ -115,7 +115,7 @@
 		}
 
 		$user = posix_getuid();
-		$configUser = fileowner(OC::$configDir . 'config.php');
+		$configUser = fileowner($config->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data'));
```
Comment (Altahrim) [inline cron.php:118]:
```suggestion
		$dataDirectoryUser = fileowner($config->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data'));
```
Need to next lines too

File: cron.php [hunk 3]
```diff
@@ -115,7 +115,7 @@
...
 		if ($user !== $configUser) {
 			echo "Console has to be executed with the user that owns the file config/config.php" . PHP_EOL;
```
Comment (Altahrim) [inline cron.php:120]:
```suggestion
			echo "Console has to be executed with the user that owns the data directory" . PHP_EOL;
```

File: console.php [hunk 4]
```diff
@@ -59,14 +59,16 @@ function exceptionHandler($exception) {
...
 
+	$config = \OC::$server->getConfig();
+
 	if (!function_exists('posix_getuid')) {
 		echo "The posix extensions are required - see https://www.php.net/manual/en/book.posix.php" . PHP_EOL;
 		exit(1);
 	}
 	$user = posix_getuid();
-	$configUser = fileowner(OC::$configDir . 'config.php');
+	$configUser = fileowner($config->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data'));
```
Comment (Altahrim) [inline console.php:69]:
```suggestion
	$dataDirectoryUser = fileowner($config->getSystemValue('datadirectory', \OC::$SERVERROOT . '/data'));
```
Need to next lines too

File: cron.php (hunk 2 above)
Comment (szaimen) [reply cron.php:118]: Maybe use getSystemValueString here?

**Review by Altahrim — APPROVED** [review]
:tada:

**Comment by Dreamsorcerer** [comment]
Thanks for finishing it off, I'm a bit short on time at the moment, so was just making quick updates in the Github UI.

**Comment by szaimen** [comment]
This breaks e.g. in https://github.com/nextcloud/all-in-one/discussions/4542#discussioncomment-9275002

**Comment by Dreamsorcerer** [comment]
> This breaks e.g. in [nextcloud/all-in-one#4542 (comment)](https://github.com/nextcloud/all-in-one/discussions/4542#discussioncomment-9275002)

Why are the users different in this case? The upgrade needs write access to that directory, so seems like it'd make sense to me that it is owned by the same user.

Though thinking a little more about that check. Surely it only needs to check that we can write to the directory. I'm not clear why it checked the owner of the file in the first place...

**Comment by miccgn** [comment]
> The upgrade needs write access to that directory, so seems like it'd make sense to me that it is owned by the same user.
Having write access to a directory doesn't necessarily mean you need to be the owner. In my case, using NFS, the owner is some UID even not existant on the local www server, but nevertheless www-data gets write access.

> Though thinking a little more about that check. Surely it only needs to check that we can write to the directory. I'm not clear why it checked the owner of the file in the first place...

Exactly.

So, how can I disable that check and get OCC back to work?

**Comment by Dreamsorcerer** [comment]
> So, how can I disable that check and get OCC back to work?

Well, right now, comment out the code. That's what I've had to do for the past 3+ years while waiting for this PR to merge...

**Comment by miccgn** [comment]
> > So, how can I disable that check and get OCC back to work?
> 
> Well, right now, comment out the code. That's what I've had to do for the past 3+ years while waiting for this PR to merge...

Thanks. Just did so - in both places (cron.php and console.php).

Probably a php function like is_writable() would be more appropriate for the task here.

**Comment by szaimen** [comment]
Hi, we had to revert this pr: https://github.com/nextcloud/server/pull/45302
See https://github.com/nextcloud/server/pull/45307 on a new approach.

**Pull request merged**